### PR TITLE
Support custom gates in pickles

### LINKED ISSUE #13946: Merge custom gates to `develop`

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR enables proving custom gates in pickles, updating the wrap circuits to enable the verification for their step proofs. This contains several changes to support this, but all implemented in such a way that the circuits not using custom gates will not change.

*However* this *does* change the pickles verifier for compatibility with the updated foreign field multiplication gate, which is a breaking change that will need to land in the Berkeley hard fork.

Fixes #13946 

Related PRs
- [snarkyjs](https://github.com/o1-labs/snarkyjs/pull/1093)
- [snarkyjs-bindings](https://github.com/o1-labs/snarkyjs-bindings/pull/128)

In it's current state, this PR is still partially incomplete. Still TODO:
- [ ] tests for the lookup gate
  - this requires some routing to inject the correct lookup tables
- [ ] tests for runtime lookup tables
  - this requires some additional routing to setup the runtime table structure and inject the values for use in the proof
- [ ] the `Maybe` versions of the logic, with tests
  - these are all a straightforward modifications of the existing logic
  - the tests require a second set of pickles inductive rules, where there is an additional rule that *doesn't* use the optional gates
- [ ] integrating the tests into the unit tests
  - this is a trivial modification to the dune file.

Checklist:

- [ ] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

File: src/lib/pickles/fix_domains.mli [hunk 1]
```diff
@@ -1,5 +1,6 @@
 val domains :
-     (module Snarky_backendless.Snark_intf.Run with type field = 'field)
+     ?min_log2:int
```
Comment (rbonichon) [inline src/lib/pickles/fix_domains.mli:2]: Could we maybe document how this new argument should be used ? (and its default value to 0).

File: src/lib/pickles/impls.ml [hunk 2]
```diff
@@ -216,9 +216,28 @@ module Wrap = struct
...
+             feature_flags
+           in
+           Array.reduce_exn ~f:Plonk_types.Opt.Flag.( ||| )
+             [| range_check0
+              ; range_check1
+              ; foreign_field_mul
+              ; xor
+              ; rot
+              ; lookup
+             |] )
```
Comment (rbonichon) [inline src/lib/pickles/impls.ml:240]: [nit] Alternatively, we could forego creating a new array
```suggestion
         range_check0 |||  range_check1 ||| foreign_field_mul ||| xor ||| rot ||| lookup
```
But maybe there's some kind of error that we might be able to catch more easily in the current form?

File: src/lib/pickles/impls.mli [hunk 3]
```diff
@@ -130,7 +130,10 @@ module Wrap : sig
   end
 
   val input :
-       unit
+       feature_flags:
+         Pickles_types.Plonk_types.Opt.Flag.t
+         Pickles_types.Plonk_types.Features.t
```
Comment (rbonichon) [inline src/lib/pickles/impls.mli:135]:
```suggestion
         Pickles_types.Plonk_types.Features.options
```
Maybe the names that I introduced (`options` / `flags`) are not appropriate or misleading

File: src/lib/pickles/plonk_checks/plonk_checks.ml [hunk 4]
```diff
@@ -496,12 +496,12 @@ module Make (Shifted_value : Shifted_value.S) (Sc : Scalars.S) = struct
...
         ; zeta_to_srs_length = pow2pow (module F) zeta env.srs_length_log2
         ; perm
-        ; lookup =
+        ; joint_combiner =
             ( match joint_combiner with
             | None ->
                 Plonk_types.Opt.None
             | Some joint_combiner ->
-                Some { joint_combiner } )
+                Some joint_combiner )
```
Comment (rbonichon) [inline src/lib/pickles/plonk_checks/plonk_checks.ml:504]: We might need an `of_option` function in `Plonk_types.Opt`. Wdyt?

File: src/lib/pickles/plonk_checks/plonk_checks.ml [hunk 5]
```diff
@@ -525,11 +525,11 @@ module Make (Shifted_value : Shifted_value.S) (Sc : Scalars.S) = struct
         ; gamma = plonk.gamma
         ; zeta = plonk.zeta
         ; joint_combiner =
-            ( match plonk.lookup with
+            ( match plonk.joint_combiner with
             | Plonk_types.Opt.None ->
                 None
             | Some l | Maybe (_, l) ->
-                Some l.In_circuit.Lookup.joint_combiner )
+                Some l )
```
Comment (rbonichon) [inline src/lib/pickles/plonk_checks/plonk_checks.ml:532]: Call `Plonk_types.Opt.to_option_unsafe plonk.joint_combiner`?

File: src/lib/pickles/step_branch_data.ml [hunk 6]
```diff
@@ -161,7 +161,30 @@ let create
...
+          } =
+        actual_feature_flags
+      in
+      let total_length =
+        ( if range_check0 || range_check1 || foreign_field_mul || rot then
+          Int.pow 2 12
+        else 0 )
+        + (if xor then 3 else Int.pow 2 8)
+        + (if lookup then 0 (* TODO: Need more info. *) else 0)
+        + if runtime_tables then 0 (* TODO: Need more info. *) else 0
```
Comment (rbonichon) [inline src/lib/pickles/step_branch_data.ml:182]: How come we have parentheses in one line and not the other?  Could it be not `ocamlformat`-formatted?

File: src/lib/pickles/test/optional_custom_gates/pickles_test_optional_custom_gates.ml [hunk 7]
```diff
@@ -0,0 +1,270 @@
+open Core_kernel
+open Pickles_types
+open Pickles.Impls.Step
+
+let () = Pickles.Backend.Tick.Keypair.set_urs_info []
```
Comment (rbonichon) [inline src/lib/pickles/test/optional_custom_gates/pickles_test_optional_custom_gates.ml:5]: Tests! Nice! 🎆

File: src/lib/pickles/wrap.ml [hunk 8]
```diff
@@ -269,13 +267,12 @@ let deferred_values (type n) ~(sgs : (Backend.Tick.Curve.Affine.t, n) Vector.t)
...
-                    .In_circuit
-                    .Lookup
-                    .joint_combiner = Option.value_exn plonk0.joint_combiner
-                  } )
+          ; joint_combiner =
+              ( match plonk0.joint_combiner with
+              | None ->
+                  Plonk_types.Opt.None
+              | Some x ->
+                  Plonk_types.Opt.Some x )
```
Comment (rbonichon) [inline src/lib/pickles/wrap.ml:275]: Second use: this gives a stronger case for adding an `of_option` function.

File: src/lib/pickles_types/pcs_batch.ml [hunk 9]
```diff
@@ -89,12 +89,18 @@ let combine_split_commitments _t ~scale_and_add ~init:i ~xi (type n)
...
-          scale_and_add ~acc ~xi p )
+  let rec go = function
+    | [] ->
+        failwith "combine_split_commitments: empty"
+    | init :: comms -> (
+        match i init with
+        | None ->
+            go comms
+        | Some init ->
+            List.fold_left comms ~init ~f:(fun acc p ->
```
Comment (rbonichon) [inline src/lib/pickles_types/pcs_batch.ml:100]: This new function could behave differently than the previous one. Do we have a rationale or a test that shows why this is so?

File: src/lib/pickles/wrap_verifier.ml [hunk 10]
```diff
@@ -521,12 +649,18 @@ struct
...
                      List.to_array (Inner_curve.to_field_elements z) )
                    m )
-                ~f:(fun x -> Sponge.absorb index_sponge x) ;
+                ~f:(function
+                  | Plonk_types.Opt.None ->
+                      ()
+                  | Plonk_types.Opt.Maybe _ ->
+                      failwith "TODO"
+                  | Plonk_types.Opt.Some x ->
+                      Array.iter x ~f:(fun x -> Sponge.absorb index_sponge x) ) ;
```
Comment (rbonichon) [inline src/lib/pickles/wrap_verifier.ml:663]: [super nit] 
```suggestion
                      Array.iter x ~f:(Sponge.absorb index_sponge) ) ;
```

File: src/lib/pickles/wrap_verifier.ml [hunk 11]
```diff
@@ -521,12 +649,18 @@ struct
...
+                (index_to_field_elements
                    ~g:(fun (z : Inputs.Inner_curve.t) ->
                      List.to_array (Inner_curve.to_field_elements z) )
                    m )
-                ~f:(fun x -> Sponge.absorb index_sponge x) ;
+                ~f:(function
+                  | Plonk_types.Opt.None ->
+                      ()
+                  | Plonk_types.Opt.Maybe _ ->
+                      failwith "TODO"
```
Comment (rbonichon) [inline src/lib/pickles/wrap_verifier.ml:661]: Could we maybe have a more extended documentation than `TODO`? I'm asking for a friend who's afraid he might need to fill the TODO in the future :-D

File: src/lib/pickles/wrap_verifier.ml [hunk 12]
```diff
@@ -617,8 +751,105 @@ struct
         absorb sponge PC (Boolean.true_, x_hat) ;
         let w_comm = messages.w_comm in
         Vector.iter ~f:absorb_g w_comm ;
+        let joint_combiner =
+          match messages.lookup with
+          | None ->
+              Types.Opt.None
+          | Maybe (b, l) ->
+              failwith "TODO"
```
Comment (rbonichon) [inline src/lib/pickles/wrap_verifier.ml:759]: This `TODO` guy again. Hope you are substituted by a more thorough string very shortly :-)

File: src/lib/pickles/wrap_verifier.ml [hunk 13]
```diff
@@ -617,8 +751,105 @@ struct
...
+                      let z = Array.map z ~f:(fun z -> (Boolean.true_, z)) in
+                      absorb sponge Without_degree_bound z ) ;
+                  Types.Opt.Some joint_combiner
+              | _ :: None :: _, None ->
+                  Array.iter l.sorted ~f:(fun z ->
+                      let z = Array.map z ~f:(fun z -> (Boolean.true_, z)) in
+                      absorb sponge Without_degree_bound z ) ;
+                  Types.Opt.Some { inner = Field.zero }
+              | _ ->
+                  failwith "TODO" )
```
Comment (rbonichon) [inline src/lib/pickles/wrap_verifier.ml:774]: You keep coming back. It's hard to let you go but we need to.

File: src/lib/pickles/wrap_verifier.ml [hunk 14]
```diff
@@ -617,8 +751,105 @@ struct
...
+        in
+        let lookup_table_comm =
+          match (messages.lookup, joint_combiner) with
+          | Types.Opt.None, Types.Opt.None ->
+              Types.Opt.None
+          | ( Types.Opt.Maybe (b_l, l)
+            , Types.Opt.Maybe (b_joint_combiner, joint_combiner) ) ->
+              failwith "TODO"
```
Comment (rbonichon) [inline src/lib/pickles/wrap_verifier.ml:782]: This one was hidden in the weeds. Found you!

File: src/lib/pickles/wrap_verifier.ml [hunk 15]
```diff
@@ -617,8 +751,105 @@ struct
...
+              in
+              let table_ids =
+                Types.Opt.map m.lookup_table_ids ~f:(fun x -> [| x |])
+              in
+              Vector.fold ~init:table_ids rest_rev ~f:(fun acc comm ->
+                  match acc with
+                  | Types.Opt.None ->
+                      comm
+                  | Types.Opt.Maybe (b, acc) ->
+                      failwith "TODO"
```
Comment (rbonichon) [inline src/lib/pickles/wrap_verifier.ml:810]: What? again?

File: src/lib/pickles/wrap_verifier.ml [hunk 16]
```diff
@@ -617,8 +751,105 @@ struct
...
+                Vector.map
+                  ~f:(Types.Opt.map ~f:(fun x -> [| x |]))
+                  m.lookup_table_comm
+              in
+              let second_column_with_runtime =
+                match (second_column, l.runtime) with
+                | Types.Opt.None, comm | comm, Types.Opt.None ->
+                    comm
+                | Types.Opt.Maybe _, _ | _, Types.Opt.Maybe _ ->
+                    failwith "TODO"
```
Comment (rbonichon) [inline src/lib/pickles/wrap_verifier.ml:794]: Woot?

File: src/lib/pickles/wrap_verifier.ml [hunk 17]
```diff
@@ -617,8 +751,105 @@ struct
...
+                  | Types.Opt.None ->
+                      comm
+                  | Types.Opt.Maybe (b, acc) ->
+                      failwith "TODO"
+                  | Types.Opt.Some acc -> (
+                      match comm with
+                      | Types.Opt.None ->
+                          Types.Opt.Some acc
+                      | Types.Opt.Maybe (b, comm) ->
+                          failwith "TODO"
```
Comment (rbonichon) [inline src/lib/pickles/wrap_verifier.ml:816]: TODO: expand the TODOs

File: src/lib/pickles/wrap_verifier.ml [hunk 18]
```diff
@@ -617,8 +751,105 @@ struct
...
+              assert false
+        in
+        let lookup_sorted =
+          (* FIXME *)
+          Vector.init Plonk_types.Lookup_sorted.n ~f:(fun i ->
+              match messages.lookup with
+              | Types.Opt.None ->
+                  Types.Opt.None
+              | Types.Opt.Maybe _ ->
+                  failwith "TODO"
```
Comment (rbonichon) [inline src/lib/pickles/wrap_verifier.ml:835]: :bomb:

File: src/lib/pickles/wrap_verifier.ml [hunk 19]
```diff
@@ -617,8 +751,105 @@ struct
...
+                  try Types.Opt.Some l.sorted.(i) with _ -> Types.Opt.None ) )
+        in
         let beta = sample () in
         let gamma = sample () in
+        let () =
+          match messages.lookup with
+          | None ->
+              ()
+          | Maybe (b, l) ->
+              failwith "TODO"
```
Comment (rbonichon) [inline src/lib/pickles/wrap_verifier.ml:846]: 💥

File: src/lib/pickles/wrap_verifier.ml [hunk 20]
```diff
@@ -673,8 +907,25 @@ struct
              It should be sufficient to fork the sponge after squeezing beta_3 and then to absorb
              the combined inner product.
           *)
-          let num_commitments_without_degree_bound = Nat.N45.n in
+          let len_1, len_1_add = Plonk_types.(Columns.add Permuts_minus_1.n) in
+          let len_2, len_2_add = Plonk_types.(Columns.add len_1) in
+          let _len_3, len_3_add = Nat.N9.add len_2 in
+          let _len_4, len_4_add = Nat.N6.add Plonk_types.Lookup_sorted.n in
+          let len_5, len_5_add =
+            (* NB: Using explicit 11 because we can't get add on len_4 *)
```
Comment (rbonichon) [inline src/lib/pickles/wrap_verifier.ml:915]: That's somewhat sad but that's commented. Bonus points!

File: src/lib/pickles/wrap_verifier.ml [hunk 21]
```diff
@@ -673,8 +907,25 @@ struct
...
+          let num_commitments_without_degree_bound = len_6 in
           let without_degree_bound =
+            let append_chain len second first =
+              Vector.append first second len
+            in
+            let undo_chunking =
+              Types.Opt.map ~f:(fun x ->
+                  assert (Array.length x = 1) ;
+                  x.(0) )
+            in
```
Comment (rbonichon) [inline src/lib/pickles/wrap_verifier.ml:928]: Liked this function ❤️

**Review by rbonichon — APPROVED** [review]
I'm approving because this is mostly good and my comments are not far away from nits, or are entirely nits. 

I do feel we should address the numerous TODOs (as stated in the PR description). This could be a general comment within the file where they occur if they have to stay as TODOs.

Besides. it should be clear that the `go` function that changed is correct (I guess it is, but let's double-check)

File: src/lib/crypto/kimchi_bindings/stubs/src/pasta_fq_plonk_verifier_index.rs [hunk 22]
```diff
@@ -76,21 +83,27 @@ impl From<CamlPastaFqPlonkVerifierIndex> for VerifierIndex<Pallas> {
...
-            lookup_features: LookupFeatures {
-                patterns: LookupPatterns {
-                    xor: false,
-                    lookup: false,
-                    range_check: false,
-                    foreign_field_mul: false,
-                },
-                joint_lookup_used: false,
-                uses_runtime_tables: false,
+            range_check0: evals.range_check0_comm.is_some(),
```
Comment (dannywillems) [inline src/lib/crypto/kimchi_bindings/stubs/src/pasta_fq_plonk_verifier_index.rs:86]: Bindings must be changed accordingly.

Comment (dannywillems) [reply src/lib/crypto/kimchi_bindings/stubs/src/pasta_fq_plonk_verifier_index.rs:86]: Opened https://github.com/MinaProtocol/mina/issues/13872.

File: src/lib/pickles/step_branch_data.ml [hunk 23]
```diff
@@ -157,10 +157,33 @@ let create
...
+          ; runtime_tables
+          } =
+        actual_feature_flags
+      in
+      let total_length =
+        ( if range_check0 || range_check1 || foreign_field_mul || rot then
+          Int.pow 2 12
+        else 0 )
+        + (if xor then 3 else Int.pow 2 8)
+        + (if lookup then 0 (* TODO: Need more info. *) else 0)
```
Comment (dannywillems) [inline src/lib/pickles/step_branch_data.ml:180]: I open [this issue](https://github.com/MinaProtocol/mina/issues/13977) to track this TODO. It is important for the use of lookup and runtime tables. I would suggest to add a link to the issue next to the TODO. This way, the reader can be redirected to a conversion outside the code if interested.

File: src/lib/pickles/step_branch_data.ml [hunk 24]
```diff
@@ -157,10 +157,33 @@ let create
...
+          ; foreign_field_mul
+          ; xor
+          ; rot
+          ; lookup
+          ; runtime_tables
+          } =
+        actual_feature_flags
+      in
+      let total_length =
+        ( if range_check0 || range_check1 || foreign_field_mul || rot then
```
Comment (dannywillems) [inline src/lib/pickles/step_branch_data.ml:176]: Can you explain the different values? I don't understand why it is 3 when xor is used and 2^8 when not used, where it is 2^12 and 0 for the other.

File: src/lib/pickles_types/plonk_types.ml [hunk 25]
```diff
@@ -1284,10 +1327,28 @@ module Messages = struct
...
     let open Snarky_backendless.Typ in
+    let uses_lookup, lookups_per_row_4 =
+      let { Features.range_check0
+          ; range_check1
+          ; foreign_field_add = _ (* Doesn't use lookup *)
+          ; foreign_field_mul
+          ; xor
+          ; rot
+          ; lookup
+          ; runtime_tables = _ (* Fixme *)
```
Comment (dannywillems) [inline src/lib/pickles_types/plonk_types.ml:1342]: A wild fixme here

**Pull request closed without merging**